streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -374,50 +379,68 @@ void handleLostAll() {
-
-        final Set<TaskId> locallyStoredTasks = new HashSet<>();
-
         final File[] stateDirs = stateDirectory.listTaskDirectories();
         if (stateDirs != null) {
             for (final File dir : stateDirs) {
                 try {
                     final TaskId id = TaskId.parse(dir.getName());
-                    // if the checkpoint file exists, the state is valid.
-                    if (new File(dir, StateManagerUtil.CHECKPOINT_FILE_NAME).exists()) {
-                        locallyStoredTasks.add(id);
+                    final Task task = tasks.get(id);
+                    if (task != null) {
+                        if (task.isActive() && task.state() == RUNNING) {
+                            taskOffsetSums.put(id, Task.LATEST_OFFSET);
+                        } else {
+                            taskOffsetSums.put(id, computeOffsetSum(task.changelogOffsets()));
+                        }
+                    } else {
+                        try {
+                            // if we are able to lock this task dir and find a valid checkpoint file, we are
+                            // responsible for encoding its offsets in our subscription
+                            final File checkpointFile = new File(dir, StateManagerUtil.CHECKPOINT_FILE_NAME);
+                            if (stateDirectory.lock(id) && checkpointFile.exists()) {
+                                taskOffsetSums.put(id, computeOffsetSum(new OffsetCheckpoint(checkpointFile).read()));
+                                lockedUnassignedTaskDirectories.add(id);
+                            }
+                        } catch (final IOException e) {
+                            // if for any reason we can't lock this task dir and read its checkpoint file, just move on
+                        }
                     }
                 } catch (final TaskIdFormatException e) {
-                    // there may be some unknown files that sits in the same directory,
-                    // we should ignore these files instead trying to delete them as well
+                    // we should just ignore any unknown files that sit in the same directory
                 }
             }
         }
 
-        return locallyStoredTasks;
+        return taskOffsetSums;
+    }
+
+    private long computeOffsetSum(final Map<TopicPartition, Long> changelogOffsets) {
+        long offsetSum = 0;
+        for (final long offset : changelogOffsets.values()) {
+            offsetSum += offset;
+        }
+        return offsetSum;
+    }
+
+    private void releaseTemporarilyLockedTaskDirectories() {
+        for (final TaskId id : lockedUnassignedTaskDirectories) {
+            try {
+                stateDirectory.unlock(id);
+            } catch (final IOException e) {
+                log.error("Failed to release the state directory lock for task {}.", id);
+                throw new StreamsException("Unable to unlock task directory after rebalance.");
```

Comment:
Should we try to unlock the rest? Also, we should always include the causing exception.